src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs

```diff
@@ -0,0 +1,111 @@
+    using Microsoft.Azure.Cosmos.Table.Queryable;
+
+    public class ASQEndpointMigrator : ITimeoutsTarget.IEndpointTargetBatchMigrator
+    {
+        public ASQEndpointMigrator(CloudTableClient client, string destinationTimeoutTableName, string stagingTableName)
+        {
+            this.client = client;
+            this.destinationTimeoutTableName = destinationTimeoutTableName;
+            this.stagingTableName = stagingTableName;
+        }
+
+        public async ValueTask<int> StageBatch(IReadOnlyList<TimeoutData> timeouts, int batchNumber)
+        {
+            await PurgeStagingTable();
+
+            var stagedTimeouts = timeouts.Select(timeout => StagedDelayedMessageEntity.FromTimeoutData(timeout, batchNumber)).GroupBy(x => x.PartitionKey);
+            var table = client.GetTableReference(stagingTableName);
+
+            var timeoutsStaged = await BatchProcessRecords(table, stagedTimeouts, (operation, entity) => operation.InsertOrReplace(entity));
+            return timeoutsStaged;
+        }
+
+        public async ValueTask<int> CompleteBatch(int batchNumber)
+        {
+            var stagingTable = client.GetTableReference(stagingTableName);
+            var query = stagingTable.CreateQuery<StagedDelayedMessageEntity>().Where(x => x.PartitionKey == batchNumber.ToString()).AsTableQuery();
+            var delayedMessages = await stagingTable.ExecuteQueryAsync(query, int.MaxValue, CancellationToken.None);
+
+            var destinationTable = client.GetTableReference(destinationTimeoutTableName);
+            var timeouts = delayedMessages.Select(DelayedMessageEntity.FromStagedTimeout);
+
+            var groupedTimeouts = timeouts.GroupBy(timeout => timeout.PartitionKey);
+            var timeoutsMigrated = await BatchProcessRecords(destinationTable, groupedTimeouts, (operation, entity) => operation.InsertOrReplace(entity));
+
+            await ClearStagedTimeoutsForBatch(batchNumber);
+            return timeoutsMigrated;
+        }
+
+        public ValueTask DisposeAsync()
+        {
+            return new ValueTask();
+        }
+
+        async Task ClearStagedTimeoutsForBatch(int batchNumber)
+        {
+            var stagingTable = client.GetTableReference(stagingTableName);
+            var query = stagingTable.CreateQuery<StagedDelayedMessageEntity>().Where(x => x.PartitionKey == batchNumber.ToString()).AsTableQuery();
+            var delayedMessages = await stagingTable.ExecuteQueryAsync(query, int.MaxValue, CancellationToken.None);
+
+            var groupedRecords = delayedMessages.GroupBy(x => x.PartitionKey);
+            await BatchProcessRecords(stagingTable, groupedRecords, (operation, entity) => operation.Delete(entity));
+        }
+
+        async Task PurgeStagingTable()
+        {
+            var table = client.GetTableReference(stagingTableName);
+            var delayedMessages = await table.ExecuteQueryAsync(new TableQuery<DelayedMessageEntity>(), int.MaxValue, CancellationToken.None);
```

Comment:
This has the drawback of materializing ALL the data from the entity to the client while all you really need to delete the entity would be the partition and rowkey